crates/containerd-shim-wasm/src/sandbox/shim/otel.rs

```diff
@@ -0,0 +1,144 @@
+//! OpenTelemetry Configuration Module
+//!
+//! This module provides a configuration structure and associated methods to initialize
+//! OpenTelemetry tracing with the OTLP exporter. The configuration can be set up via
+//! the `OtelConfig` struct and its builder pattern.
+//!
+//! # Usage
+//!
+//! ```rust
+//! use containerd-shim-wasm::sandbox::shim::otel::{OtelConfig, OTEL_EXPORTER_OTLP_ENDPOINT};
+//!
+//! fn main() -> anyhow::Result<()> {
+//!     let otel_endpoint = std::env::var(OTEL_EXPORTER_OTLP_ENDPOINT).expect("OTEL_EXPORTER_OTLP_ENDPOINT not set");
+//!     let otel_config = OtelConfig::builder()
+//!         .otel_endpoint(otel_endpoint)
+//!         .name("my-service".to_string())
+//!         .build()?;
+//!
+//!     let _guard = otel_config.init()?;
+//!
+//!     // Your application code here
+//!
+//!     Ok(())
+//! }
+//! ```
+
+use std::collections::HashMap;
+
+use opentelemetry::global::{self, set_text_map_propagator};
+use opentelemetry::trace::TraceError;
+use opentelemetry::KeyValue;
+use opentelemetry_otlp::WithExportConfig;
+use opentelemetry_sdk::propagation::TraceContextPropagator;
+use opentelemetry_sdk::{runtime, trace as sdktrace, Resource};
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt as _;
+use tracing_subscriber::layer::SubscriberExt as _;
+use tracing_subscriber::{EnvFilter, Registry};
+
+pub const OTEL_EXPORTER_OTLP_ENDPOINT: &str = "OTEL_EXPORTER_OTLP_ENDPOINT";
+
+/// Configuration struct for OpenTelemetry setup.
+pub struct OtelConfig {
+    otel_endpoint: String,
+    name: String,
+}
+
+impl OtelConfig {
+    /// Creates a new builder for `OtelConfig`.
+    pub fn builder() -> OtelConfigBuilder {
+        OtelConfigBuilder::default()
+    }
+
+    /// Initializes a new OpenTelemetry tracer with the OTLP exporter.
+    ///
+    /// Returns a `Result` containing the initialized tracer or a `TraceError` if initialization fails.
+    ///
+    /// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options
+    fn init_tracer(&self) -> Result<opentelemetry_sdk::trace::Tracer, TraceError> {
+        opentelemetry_otlp::new_pipeline()
+            .tracing()
+            .with_exporter(
+                opentelemetry_otlp::new_exporter()
+                    .tonic()
+                    .with_endpoint(&self.otel_endpoint),
+            )
+            .with_trace_config(sdktrace::config().with_resource(Resource::new(vec![
+                KeyValue::new("service.name", format!("containerd-shim-{}", self.name)),
```

Comment:
This should be taken from env (which I expect the otel libs to do by default).
I guess it should be ok to specify a name but then we should look at that before applying the config.
Also not sure we should prepend `containerd-shim-` in that case.

Reply:
Good point, it looks like the library will check for `OTEL_SERVICE_NAME` first

Reply:
Reverted this change to use the default service.name from OTEL_SERVICE_NAME